[#24220] Supports Supervision in Partition stage

yoshiyoshifujii opened the pull request:
#24220

yoshiyoshifujii committed:
[#24220] Supports Supervision in Partition stage

lightbend-cla-validator commented:
Hi @yoshiyoshifujii,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

ktoso commented:
OK TO TEST

ktoso commented:
From tokyo hackathon, thanks @yoshiyoshifujii !

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala:
```diff
@@ -753,27 +754,39 @@ final class Partition[T](val outputPorts: Int, val partitioner: T ⇒ Int, val e
   override val shape: UniformFanOutShape[T, T] = UniformFanOutShape[T, T](in, out: _*)
```
As mentioned during hackathon, this needs to extend the documentation with information that `* Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.` (exact that wording), in the docs on the class ` final class Partition[T]`

yoshiyoshifujii committed:
added test and documentation

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphPartitionSpec.scala:
```diff
@@ -42,6 +42,49 @@ class GraphPartitionSpec extends StreamSpec {
+        val partition = b.add(Partition[Int](3, {
+          case g if (g > 3)  ⇒ 0
+          case l if (l < 3)  ⇒ 1
+          case e if (e == 3) ⇒ throw new RuntimeException()
```
use `akka.stream.testkit.Utils.TE` to avoid scary stack traces in log output

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphPartitionSpec.scala:
```diff
@@ -42,6 +42,49 @@ class GraphPartitionSpec extends StreamSpec {
+        val partition = b.add(Partition[Int](3, {
+          case g if (g > 3)  ⇒ 0
+          case l if (l < 3)  ⇒ 1
+          case e if (e == 3) ⇒ throw new RuntimeException()
```
^

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala:
```diff
@@ -753,27 +756,39 @@ final class Partition[T](val outputPorts: Int, val partitioner: T ⇒ Int, val e
+      try {
+        val idx = partitioner(elem)
+        if (idx < 0 || idx >= outputPorts) {
+          failStage(PartitionOutOfBoundsException(s"partitioner must return an index in the range [0,${outputPorts - 1}]. returned: [$idx] for input [${elem.getClass.getName}]."))
```
should we support supervision for this also? then it should throw instead of `failStage`

ktoso replied:
This should be changed to throwing @yoshiyoshifujii

chbatey replied:
@yoshiyoshifujii I think this one was missed. We want to be able to supervise `PartitionOutOfBoundsException`. Now this isn't in the throw it could just call the decide directly.

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala:
```diff
@@ -753,27 +756,39 @@ final class Partition[T](val outputPorts: Int, val partitioner: T ⇒ Int, val e
+          case Supervision.Stop ⇒
+            failStage(ex)
+          case Supervision.Restart ⇒
+            pull(in)
```
I think this is somewhat scary since it's not pulling in all normal cases. Perhaps we should only have the try-catch around the `partitioner` invocation?

ktoso replied:
Yeah maybe good idea to move it just around that 👍

patriknw reviewed (changes requested):
looking good, some feedback...

yoshiyoshifujii committed:
Merge branch 'master' of https://github.com/akka/akka into wip-24220

yoshiyoshifujii committed:
use akka.stream.testkit.Utils.TE

yoshiyoshifujii committed:
only have the try-catch around the partitioner

yoshiyoshifujii commented:
@patriknw Fixed. Please review again.

johanandren reviewed (commented):
Two minor comments away from LGTM!

yoshiyoshifujii committed:
Merge branch 'master' of https://github.com/akka/akka into wip-24220

yoshiyoshifujii committed:
add lazy and old try-catch style.

https://github.com/akka/akka/pull/24740/files#r214642280
https://github.com/akka/akka/pull/24740/files#r214641856

johanandren reviewed (approved):
LGTM. Thanks!

yoshiyoshifujii commented:
@patriknw can you please review this again?

chbatey reviewed (changes requested):
@yoshiyoshifujii I think the last thing that is missing for this PR is to be able to supervise the PartitionOutOfBoundsException

patriknw commented:
Thanks for your patience here, @yoshiyoshifujii.
Please rebase and look at chbatey's comment above

patriknw commented:
I'll rebase this

patriknw commented:
Completed in https://github.com/akka/akka/pull/26731
